stdlib/public/stubs/LibcShims.cpp

```diff
@@ -348,31 +347,36 @@ SWIFT_RUNTIME_STDLIB_INTERNAL
 void swift::_stdlib_random(void *buf, __swift_size_t nbytes) {
   while (nbytes > 0) {
     __swift_ssize_t actual_nbytes = -1;
+
 #if defined(GETRANDOM_AVAILABLE)
     static const bool getrandom_available =
       !(getrandom(nullptr, 0, 0) == -1 && errno == ENOSYS);
+      
     if (getrandom_available) {
       actual_nbytes = WHILE_EINTR(getrandom(buf, nbytes, 0));
     }
 #elif defined(__Fuchsia__)
     __swift_size_t getentropy_nbytes = std::min(nbytes, __swift_size_t{256});
+    
     if (0 == getentropy(buf, getentropy_nbytes)) {
       actual_nbytes = getentropy_nbytes;
     }
 #endif
+
     if (actual_nbytes == -1) {
-      static const int fd =
-        WHILE_EINTR(_stdlib_open("/dev/urandom", O_RDONLY | O_CLOEXEC, 0));
+      int fd = _stdlib_open("/dev/urandom", O_RDONLY | O_CLOEXEC, 0);
```

Comment:
`_stdlib_open` and `_stdlib_close` can be interrupted by a signal (see the `WHILE_EINTR` macro).

Reply:
According to: https://lwn.net/Articles/576478/ retrying `close()` on `EINTR` is potentially hazardous. `open()` probably doesn't have this problem, but for reference, llvm's libcxx does something similar: https://github.com/llvm-mirror/libcxx/blob/master/src/random.cpp#L83 where they only retry on `read()`.

Reply:
Thanks, that was an interesting article. There are some [notes about the `EINTR` error](http://man7.org/linux/man-pages/man2/close.2.html#NOTES) in the Linux man-page.

The difference between the libc++ and Swift APIs is the error handling model:

* the `std::random_device` constructor throws an exception which can be caught;
* the `swift::_stdlib_random` function stops execution with a fatal error.

[Android's `getentropy_urandom` helper](https://android.googlesource.com/platform/bionic.git/+/master/libc/bionic/getentropy.cpp) retries `open` and `read`, but ignores all `close` errors.